tests/artifacts_tests/test_boto3.py

```diff
@@ -0,0 +1,85 @@
+from __future__ import annotations
+
+import io
+from typing import TYPE_CHECKING
+
+import boto3
+from moto import mock_s3
+import pytest
+
+from optuna.artifacts import Boto3ArtifactStore
+from optuna.artifacts.exceptions import ArtifactNotFound
+
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+    from mypy_boto3_s3 import S3Client
+    from typing_extensions import Annotated
+
+    # TODO(Shinichi) import Annotated from typing after python 3.8 support is dropped.
+
+
+@pytest.fixture()
+def init_mock_client() -> Iterator[tuple[str, S3Client]]:
+    with mock_s3():
+        # Runs before each test
+        bucket_name = "moto-bucket"
+        s3_client = boto3.client("s3")
+        s3_client.create_bucket(Bucket=bucket_name)
+
+        yield bucket_name, s3_client
+
+        # Runs after each test
+        objects = s3_client.list_objects(Bucket=bucket_name).get("Contents", [])
+        if objects:
+            s3_client.delete_objects(
+                Bucket=bucket_name,
+                Delete={"Objects": [{"Key": obj["Key"] for obj in objects}], "Quiet": True},
+            )
+        s3_client.delete_bucket(Bucket=bucket_name)
+
+
+def test_upload_download(
```

Comment:
How about adding the test case for `avoid_buf_copy` using `pytest.mark.parametrize`?

Reply:
I added test cases.